### PR TITLE
ZKD Indexes

### PR DESCRIPTION
**Pull request by maierlars**
### Scope & Purpose
The multi-dimensional index allows to combine multiple range queries. For each dimension you can specify an interval using `<`, `>`, `>=`, `<=`, and `==`. The interval can be bounded one both sides or only on one side. You can leave dimensions unbounded as well.

```
Query String (107 chars, cacheable: true):
 for doc in c
     Filter doc.x >= 4 && doc.x <= 8
     Filter doc.y > 15
     Filter doc.z == 8
     return doc

Execution plan:
 Id   NodeType        Est.   Comment
  1   SingletonNode      1   * ROOT
 10   IndexNode          0     - FOR doc IN c   /* zkd index scan */    FILTER (doc.`y` > 15)   /* early pruning */   
  9   ReturnNode         0       - RETURN doc

Indexes used:
 By   Name                      Type   Collection   Unique   Sparse   Selectivity   Fields              Ranges
 10   idx_1693858824336703488   zkd    c            false    false            n/a   [ `x`, `y`, `z` ]   ((doc.`y` >= 15) && (doc.`x` >= 4) && (doc.`z` == 8) && (doc.`x` <= 8))
```

- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)

TODO:

- [ ] Make some experiments if there's a limit how many dimensions work (even if not fast), or if it crashes, and if we have to limit the number of dimensions somehow.

#### Related Information
Still missing.

- [ ] Docs PR: https://github.com/arangodb/docs/pull/633
- [ ] Design document: 
- [ ] CHANGELOG

### Testing & Verification
Present but could be more.

- [x] The behavior in this PR was *manually tested*
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new C++ **Unit tests**
  - [x] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests**, **required**
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/14312/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/15601/

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16024/

**Comment by maierlars** [comment]
image

**Comment by gnusi** [comment]
I fear that the way of how `zkd` index functionality is exposed for the user is a bit counterintuitive. In order to use it, one has to generate a JSON object where fields represent a certain dimensions. That's weird, given json natively supports arrays and vpack representation of an array is by far more efficient than object.

Instead of specifying object fields to be indexed, I suggest the following:
```
{ vec: [ 0, 1, 2, 3, 4 ] }
```
index definition
```
{ 
  fields: [ "vec" ], // only 1 field is allowed
  dimensions: 4 // consider only 4 dimension, e.g. this can raise an error if "vec" contains less than 4 dimensions
}
```
or 
```
{ 
  fields: [ "vec" ], // only 1 field is allowed
  dimensions: [ 0, 1, 3 ] // index only 0, 1 and 3 dimensions
}
```

AQL will look like
```
FOR d IN c
  FILTER d.vec[0] > 0 && d.vec[0] < 42
  FILTER d.vec[1] > 2 && d.vec[1] <= 48
  FILTER d.vec[3] == 8
  RETURN d
```

**Comment by gnusi** [comment]
I also fear that `zkd` might be a bit misleading for average user, I suggest something like `multidimensional` (but probably it's too long)

**Comment by gnusi** [comment]
I also suggest we introduce a reasonable upper bound for the number of dimensions

**Comment by goedderz** [comment]
> I fear that the way of how `zkd` index functionality is exposed for a user is a bit counterintuitive. In order to use it, one has to generate a JSON object where fields represent certain dimensions. That's weird, given json natively supports arrays and vpack representation of an array is by far more efficient than object.
> 
> Instead of specifying object fields to be indexed, I suggest the following:
> 
> ```
> { vec: [ 0, 1, 2, 3, 4 ] }
> ```

That's a sensible idea for this use case, i.e. indexing homogenic dimensions. However, we (also) had in mind more heterogenic dimensions, at least in the future. While we decided to restrain this first version to doubles only, which does limit that somewhat, we have plans how to allow arbitrary json/velocypack values later. In which case allowing one one array to be indexed feels much too restricting.

> I also fear that zkd might be a bit misleading for average user, I suggest something like multidimensional (but probably it's too long)

Agreed, that's why we used multi-dimensional in the documentation. However, as other kinds of multidimensional indexes are available and could be added in the future, we thought it sensible to use the more technical term in the API. And it's also common to specify the implementation of indexes, i.e. btree, hash, rtree, skiplist, etc.

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16576/

File: arangod/RocksDBEngine/RocksDBZkdIndex.cpp [hunk 1]
```diff
@@ -0,0 +1,569 @@
...
+                                                   arangodb::velocypack::Slice doc) {
+  TRI_ASSERT(_unique == false);
+  TRI_ASSERT(_sparse == false);
+
+  auto key_value = readDocumentKey(doc, _fields);
+
+  RocksDBKey rocks_key;
+  rocks_key.constructZkdIndexValue(objectId(), key_value, documentId);
+
+  auto value = RocksDBValue::ZkdIndexValue();
```
Comment (gnusi) [inline arangod/RocksDBEngine/RocksDBZkdIndex.cpp:459]: seems to be redundant

File: arangod/RocksDBEngine/RocksDBZkdIndex.cpp [hunk 2]
```diff
@@ -0,0 +1,569 @@
...
+    arangodb::velocypack::Slice doc) {
+  TRI_ASSERT(_unique == true);
+  TRI_ASSERT(_sparse == false);
+
+  auto key_value = readDocumentKey(doc, _fields);
+
+  RocksDBKey rocks_key;
+  rocks_key.constructZkdIndexValue(objectId(), key_value);
+
+  auto value = RocksDBValue::ZkdIndexValue();
```
Comment (gnusi) [inline arangod/RocksDBEngine/RocksDBZkdIndex.cpp:562]: seems to be redundant

**Review by gnusi — CHANGES_REQUESTED** [review]
There are 2 critical things:
* usage of a rude word for a label
* lack of disclaimer in some files

Other comments are suggestions for performance improvements. 
I'd also add some load tests against at least medium size data set (probably as a part of QA phase).

**Review by mpoeter — APPROVED** [review]
LGTM, only some small suggestions.

**Comment by maierlars** [comment]
https://jenkins.arangodb.biz/job/arangodb-matrix-pr/16971/

File: arangod/Zkd/ZkdHelper.cpp [hunk 3]
```diff
@@ -0,0 +1,684 @@
...
+
+auto zkd::compareWithBox(byte_string_view cur, byte_string_view min, byte_string_view max, std::size_t dimensions)
+-> std::vector<CompareResult> {
+  if (dimensions == 0) {
+    auto msg = std::string{"dimensions argument to "};
+    msg += __func__;
+    msg += " must be greater than zero.";
+    throw std::invalid_argument{msg};
+  }
+  std::vector<CompareResult> result;
```
[the remark this answers is not in the document]
Comment (goedderz) [reply arangod/Zkd/ZkdHelper.cpp:275]: Good idea, but I'll create a ticket and do that as a later improvement.

File: arangod/RocksDBEngine/RocksDBZkdIndex.cpp [hunk 4]
```diff
@@ -0,0 +1,567 @@
...
+    doc = doc.get(attrib.name);
+  }
+
+  return doc;
+}
+
+auto readDocumentKey(VPackSlice doc,
+                     std::vector<std::vector<basics::AttributeName>> const& fields)
+    -> zkd::byte_string {
+  std::vector<zkd::byte_string> v;
```
[the remark this answers is not in the document]
Comment (goedderz) [reply arangod/RocksDBEngine/RocksDBZkdIndex.cpp:198]: Good idea, but I'll create a ticket and do that as a later improvement.

**Review by joerg84 — APPROVED** [review]

**Pull request merged**